WikiWatershed/mmw-geoprocessing
# Add RasterSummary Operation

## Overview

Adds `RasterSummary` operation, needed to generate summary statistics (min, average, and max) for a list of layers within a given shape.

In accordance with the `LongAdder` and `DoubleAdder` used for counting and averaging in `RasterGroupedAverage`, we use a `DoubleAccumulator` for both min and max values, with custom `MinWithoutNoData` and `MaxWithoutNoData` implementations that ignore `NoData` values and only count the rest.

This will only work with `Double` type layers, not with `Int` layers.

Connects #78 
Connects https://github.com/WikiWatershed/model-my-watershed/issues/2346

### Demo

```http
$ http :8090/run < examples/MapshedJob_RasterSummary.json

HTTP/1.1 200 OK
Content-Length: 124
Content-Type: application/json
Date: Fri, 29 Dec 2017 23:08:53 GMT
Server: akka-http/10.0.9

{
    "result": [
        {
            "avg": 799.9116994958924,
            "max": 3430.0,
            "min": -38.0
        },
        {
            "avg": 1.2253132788318364,
            "max": 28.67453956604004,
            "min": 0.0
        }
    ]
}
```

Thanks to @arottersman who confirmed the first set of values above by running GDAL using the same shape on the original TIFF and getting:

```
Band 1 Block=475x4 Type=Int32, ColorInterp=Gray
  Description = elev_cm
  Minimum=-38.000, Maximum=3430.000, Mean=799.912, StdDev=630.396
  NoData Value=-2147483647
  Metadata:
    LAYER_TYPE=athematic
    STATISTICS_MAXIMUM=3430
    STATISTICS_MEAN=799.91169949589
    STATISTICS_MINIMUM=-38
    STATISTICS_STDDEV=630.39608483669
```

```bash
$ ./scripts/benchmark --rs

Timing HUC8 RasterSummary ->

HUC8 RasterSummary, run 1 -> 5.618318 s
HUC8 RasterSummary, run 2 -> 4.926943 s
HUC8 RasterSummary, run 3 -> 6.105659 s
HUC8 RasterSummary, run 4 -> 4.520898 s
HUC8 RasterSummary, run 5 -> 4.682257 s

HUC8 RasterSummary average -> 5.170815 s
```

### Notes

This is the only operation that returns a result not in this format:

```json
{
  "List(1,2)": 3,
  "List(4,5)": 6
}
```

thus will need a custom parser in Django / Celery.

## Testing Instructions

* Check out this branch
* Run the example query
* Run the benchmark
* Understand the code / comment any improvements
